src/backend/avx2/scalar_mul/pippenger.rs

```diff
@@ -0,0 +1,172 @@
+
+/// Implements a version of Pippenger's algorithm.
+///
+/// See the documentation in the serial `scalar_mul::pippenger` module for more info.
+pub struct Pippenger;
+
+#[cfg(any(feature = "alloc", feature = "std"))]
+impl VartimeMultiscalarMul for Pippenger {
+    type Point = EdwardsPoint;
+
+    fn vartime_multiscalar_mul<I, J>(scalars: I, points: J) -> EdwardsPoint
+    where
+        I: IntoIterator,
+        I::Item: Borrow<Scalar>,
+        J: IntoIterator,
+        J::Item: Borrow<EdwardsPoint>,
+    {
+        let mut scalars = scalars.into_iter();
+        let size = scalars.by_ref().size_hint().0;
+        let w = if size < 3 {
+            1
+        } else if size < 13 {
+            2
+        } else if size < 27 {
+            3
+        } else if size < 66 {
+            4
+        } else if size < 190 {
+            5
+        } else if size < 500 {
+            6
+        } else if size < 800 {
+            7
+        } else if size < 3_000 {
+            8
+        } else if size < 6_000 {
+            9
+        } else if size < 13_000 {
+            10
+        } else if size < 23_000 {
+            11
+        } else if size < 45_000 {
+            12
+        } else if size < 160_000 {
+            13
+        } else if size < 260_000 {
+            14
+        } else {
+            15
+        };
+
+        let window_size = 1 << w;
+        let windows_count = (256 + w - 1) / w; // == ceil(256/w)
+        let buckets_count = window_size / 2; // digits are signed+centered hence 2^w/2, excluding 0-th bucket
+
+        // Step 1. Prepare scalar representation using signed digits radix 2^w.
+        // Each digit will be in [-2^w/2, 2^w/2].
+        let all_digits: Vec<_> = scalars
+            .into_iter()
+            .map(|c| c.borrow().to_arbitrary_radix(w))
+            .collect();
+        let ps: Vec<CachedPoint> = points
+            .into_iter()
+            .map(|p| CachedPoint::from(ExtendedPoint::from(*p.borrow())))
+            .collect();
+
+        // Iterate the digits from last to first so we can shift the result by w bits
+        // at the end of each iteration (e.g. w doublings).
+        let mut result = ExtendedPoint::identity();
+        for i in (0..windows_count).rev() {
+            // Step 2. Shift the current result by w bits to leave room for a new window.
+            if i < (windows_count - 1) {
+                result = result.mul_by_pow_2(w as u32);
+            }
+
+            // Step 3. Prepare 2^w/2 buckets.
+            // buckets[i] corresponds to a multiplication factor (i+1).
+            let mut buckets = vec![ExtendedPoint::identity(); buckets_count];
+
+            // Step 4. Iterate over pairs of (point, scalar)
+            // and add/sub the point to the corresponding bucket.
+            // Note: when we add support for precomputed lookup tables,
+            // we'll be adding/subtractiong point premultiplied by `digits[i]` to buckets[0].
```

Comment:
Note 2: adding precomputed points to buckets[0] won't really help us with performance if we have at least one non-precomputed point. This is because we'd have to add up all buckets anyway.